### ISSUE TITLE
Local Directory with User Annotated Genomes

### ISSUE THREAD

**Issue by ak-watson** [GCA-VH-lab/FlaGs#4]
I was looking to use FlaGs with some user annotated genomes using the -l and -ld commands (using prokka). 

I noticed that the way FlaGs parses GFF attributes seems to be hard-coded in so that local files would only really work with gff3 files from NCBI.

For example:  FlaGs looks for the "Name=" field as the 4th entry in the attributes section of the gff3 file, while in my prokka output (including the "genbank compliant" mode) this is the second entry. 

`if Line[8].split(';')[3][:5]=='Name=':`

I'm not sure if that is the intent, but it might be useful to highlight it in the manual.

If the intent was that it could also be used with user annotated genomes, then looking at the code, it probably wouldn't be difficult to make parts like that more generalised so that gff/faa files from different sources could be used more easily (e.g. by reading attributes into a dictionary with attribute tags as keys and the data as values, and checking for "Name" in the attribute dictionary).

Or otherwise, it might be useful to highlight which gff3 attribute fields are important/essential for FlaGs in the manual. For example. Which gff3 attribute is used to map a CDS entry to the corresponding protein sequence in the .faa file. Or highlighting that coding sequences without the "Name" attribute are filtered out (to remove potential pseudogenes).  

At the moment I am going through the code to figure out which gff3 elements are hardcoded so that I can edit mine accordingly. 

The filter based on the "Name" field is actually a little problematic for me anyway. I understand the reasoning (to filter out pseudogenes) but since I am interested in a conserved hypothetical protein, there is no "Name" field in my GFF so I guess that it would be excluded from the dataset?

Thanks,
Andrew

**Comment by ak-watson** [comment]
Just as an update, I made a slightly edited version of FlaGs_linux.py and everything is working nicely with my prokka annotated genomes.

It is still hard-coded to work for my annotation, so I can't guarantee it would work for other prokka annotated genomes, but I'm  happy to share it if it is useful. 

I did make some other small changes, like removing the requirement of the "Name" field and using the unique ID to map to the correct genes rather than the name.

Thanks,
Andrew

**Comment by frihaka** [comment]
Hi ak-watson,

I am having a similar issue as you described above: _**FlaGs_linux.py fails with non-NCBI gff files**_ (from prokka 1.14.5), most likely due to keyword such as "Name=".

I have tried to trick by playing around with my gff file to "mimic" an NCBI-gff, with no luck so far.

Would you mind please sharing your edited version?

Thanks

**Comment by frihaka** [comment]
I have tested FlaGs_linux.py in **_local mode_**, with both types of gff files/queries: ncbi-gff and (user) prokka-gff.
The first one (ncbi-gff) runs with no error and outputs the expected results.
The second one (prokka-gff) finishes but outputs "Flanking Genes Not Found".

Looking more closely to the gff files format (ncbi versus prokka) and the FlaGs_linux.py code, my own issue does not seem to be caused by the 'Name=' field's index, as suggested by @ak-watson:
> For example: FlaGs looks for the "Name=" field as the 4th entry in the attributes section of the gff3 file, while in my prokka output (including the "genbank compliant" mode) this is the second entry.
> 
> `if Line[8].split(';')[3][:5]=='Name=':`

In both NCBI-gff and prokka-gff, the indices for 'Name=' field are the same. The index value depends on the **gene** (where Name= is 2nd field) or **CDS** (where Name= is 4th field) feature of a given locus-tag.
And the code (line 976) is selecting only the CDS row:
`if Line[2]=='CDS':`

The prokka-gff writes in the CDS row's 4th field "Name=gene name" (e.g Name=rsmA), instead of "Name=query ID" (as does the NCBI-gff). This is a problem, since the code  (line 981) is looking for the exact query value matched in the Name field:
`if Line[8].split(';')[3].split('=')[1]==query.split('#')[0]:`
 
Therefore I have edited the prokka-gff file at the CDS rows in order to comply with this match (Name=queryID).

Unfortunately, the edited prokka-gff (now compliant with the code's requirements, dictionnaries derived from CDS rows seem ok, dictionaries derived from gene rows seem ok) stills leads to a "Flanking Genes Not Found". This seems to correspond to the case of line 1321:
```
								else:
									if query not in FoundDict:
										FlankFoundDict[query]='No'
										if args.verbose:
											print('\t', query.split('#')[0], 'Report: Flanking Genes Not Found', '\n')
```

If you have an idea of what's going on?

Cheers

**Comment by ak-watson** [comment]
> I have tested FlaGs_linux.py in **_local mode_**, with both types of gff files/queries: ncbi-gff and (user) prokka-gff.
> The first one (ncbi-gff) runs with no error and outputs the expected results.
> The second one (prokka-gff) finishes but outputs "Flanking Genes Not Found".
> 
> Looking more closely to the gff files format (ncbi versus prokka) and the FlaGs_linux.py code, my own issue does not seem to be caused by the 'Name=' field's index, as suggested by @ak-watson:
> 
> > For example: FlaGs looks for the "Name=" field as the 4th entry in the attributes section of the gff3 file, while in my prokka output (including the "genbank compliant" mode) this is the second entry.
> > `if Line[8].split(';')[3][:5]=='Name=':`
>
> In both NCBI-gff and prokka-gff, the indices for 'Name=' field are the same. The index value depends on the **gene** (where Name= is 2nd field) or **CDS** (where Name= is 4th field) feature of a given locus-tag.
> And the code (line 976) is selecting only the CDS row:
> `if Line[2]=='CDS':`
> 
> The prokka-gff writes in the CDS row's 4th field "Name=gene name" (e.g Name=rsmA), instead of "Name=query ID" (as does the NCBI-gff). This is a problem, since the code (line 981) is looking for the exact query value matched in the Name field:
> `if Line[8].split(';')[3].split('=')[1]==query.split('#')[0]:`
> 
> Therefore I have edited the prokka-gff file at the CDS rows in order to comply with this match (Name=queryID).
> 
> Unfortunately, the edited prokka-gff (now compliant with the code's requirements, dictionnaries derived from CDS rows seem 

Can you quickly double-check your CDS gff row. In mine (also using compliant mode), the "Name" flag is the third tag in the list, not the fourth. 

For me, I didn't want the "Name" check so I removed it completely.

Then, rather than edit my gff file, for the exact mapping later I just pointed to the correct field in the prokka output. 

My queries all match the "ID=" tag, the first field. 

So line 982 becomes: 

`if Line[8].split(';')[0].split('=')[1]==query.split('#')[0]:`

This will hopefully get you to the point where your queries are found, but there are also more changes to make later. 

I've linked to my fork of the repo. You can try the script FlaGs_linux_prokka_test.py. 

As an advanced warning just to remind you, this removes the filter for the "Name" field so will include CDs without an assigned "Name". There is also one little error in there with how I handled the genDes field (that was missing in my prokka output) but it shouldn't have too much impact on the results.

https://github.com/ak-watson/FlaGs

There are a few other minor changes as well. There is a relatively messy output I generate from PIL as a test for a simple way to work around a problem of the edges of an image being cut off to the size of the window in tkinter for a large number of flanking genes. I didn't get around to trying anything better just yet.

Thanks,
Andrew

**Comment by frihaka** [comment]
Hi,

thanks a lot for your feedback.

I had a check again on the prokka-gff. Well actually, the index of the Name field, in the CDS row, is variable...:
ID - Parent - Name
ID - Parent - eC_number - Name

(and sometimes the Name field is even absent.) 

"This" would explain "that"...


Cheers

ps: for comparison, in ncbi-gff, the Name field's index seems to be constant:
ID - Parent - Dbxref - Name

**Comment by frihaka** [comment]
I have tried your modified prokka-compatible script: it works on my prokka-gff.

**Comment by ak-watson** [comment]
No problem.

I'm glad it worked on the prokka-gff.

I used the ID and parent fields because they were the ones that were always the same in prokka. I just ignore the name field. 

To make it work with NCBI and prokka wouldn't be too difficult.

I was kind of hurrying and didn't want to spend too much time on it.

But to make something work with both, I would probably read all attributes into a dictionary where the tag is a key and the value is the value for that tag.

If "attributes" is the attributes string then something like this:

attr_dict = dict(i.split("=") for i in attributes.split(";"))

Then you could alter the code to reference the specific tag rather than the position.

E.g. 

if "Name" in attr_dict: #To check whether the name field exists 

This should work across both formats.

**Comment by frihaka** [comment]
Indeed, it would be helpful to have a script working with targets from both types of gff files. 
Would make sense to have such a dict, instead of an index-based lookup.

Thanks a lot for your quick fixes and tips!
